src/compiler/checker.ts

```diff
@@ -6112,6 +6135,18 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 approximateLength: 0,
                 trackedSymbols: undefined,
                 bundled: !!compilerOptions.outFile && !!enclosingDeclaration && isExternalOrCommonJsModule(getSourceFileOfNode(enclosingDeclaration)),
+                isEntityNameVisible: (entityName, shouldComputeAliasToMakeVisible) => isEntityNameVisible(entityName, context.enclosingDeclaration!, shouldComputeAliasToMakeVisible),
+                isExpandoFunctionDeclaration,
+                isOptionalParameter,
+                isLiteralComputedName,
+                trackComputedName(accessExpression) {
+                    trackComputedName(accessExpression, context.enclosingDeclaration, context);
+                },
+                getAllAccessorDeclarations: getAllAccessorDeclarationsForDeclaration,
+                requiresAddingImplicitUndefined,
+                isUndefinedIdentifier(node: Identifier) {
+                    return getResolvedSymbol(node) === undefinedSymbol;

```

Comment:
`getResolvedSymbol(node)` on arbitrary identifiers is spooky - it's only safe to call on specific kinds of expression-context identifiers (and if you call it on an `undefined` identifier in a type context, it's going to do wonky things, like resolve `undefined` in that scope as a value). `getSymbolAtLocation(node)` is the general-purpose version that figures out the right symbol lookup method from the context. Since you're only using this on expression-identifiers (and not type references or declaration names), maybe renaming to `isUndefinedIdentifierExpression` would be more clear? A `Debug.assert(isExpressionNode(node))` would not be unwarranted.

Reply:
I was going by what happens in `isNullOrUndefined` but that is indeed called in an expression context. Will rename it,